### PR TITLE
feature: timed bounty action added for timed bounty labels

### LINKED ISSUE #3883: Have the ability to set a timed bounty on an issue
Variable for hours.

### PR DESCRIPTION
**Pull request by swaparup36**
Resolves #3883 

Current implementation of timed bounty - if an issue is labeled like '$5 24' then it means that the bounty amount on the issue is $5 and the PR for it has to be the raised within 24 hours of the time when the label was attached to the issue. When an issue is labeled such then the action hits a webhook that will update the bounty expiry time on the database. When a pr is merged that is related to an issue that has a timed bounty then the bounty payout webhook implementation checks the time of expiry of the bounty and decides if the bounty has to be paid or not.

## Summary by CodeRabbit

* **New Features**
  * Timed bounties with expiry recording, validation, and automated handling.
  * New workflow to detect timed-bounty labels and notify the backend.
  * GitHub Sponsors payment option and automated issue comments/labels on payout.

* **Bug Fixes / Improvements**
  * Stronger expiry checks, clearer error responses, rate limiting, and improved logging.
  * Payout flow hardened to prevent duplicates and handle timed-bounty scenarios; timed-bounty flag propagated through the workflow.

* **Tests**
  * Added comprehensive tests covering timed-bounty endpoints and payout edge cases.

### CODE REVIEW

**Review by Nachiket-Roy — CHANGES_REQUESTED** [review]
1. Missing workflow permission restrictions.
2. Add Rate-limit
3. Proper validation for the bounty duration.

**Comment by swaparup36** [comment]
> 1. Missing workflow permission restrictions.
> 2. Add Rate-limit
> 3. Proper validation for the bounty duration.

Can you please elaborate the points? what I should change exactly?

**Review by Nachiket-Roy — COMMENTED** [review]
@coderabbitai Please verify whether this PR fully addresses the following items or if any are still missing:
Missing permissions in the GitHub Action
Inconsistent label parsing logic
Clean merge migration files for better history maintainability
Potential race condition

**Review by DonnieBLT — DISMISSED** [review]
please update to one migration file

**Comment by swaparup36** [comment]
> please update to one migration file 

Sure

**Review by Nachiket-Roy — CHANGES_REQUESTED** [review]
Sorry for the late response. I’ll re-review once you squash the migration and fix the unresolved comment. Please request a re-review after that, and feel free to drop a message in the Slack channel as well (if i miss again)

File: .github/workflows/timed-bounty-label.yml [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+              body: JSON.stringify(payload)
+            });
+
+            if (!response.ok) {
+              const errorText = await response.text();
+              throw new Error(`Timed bounty update failed: ${response.status} - ${errorText}`);
+            }
+
+            const result = await response.json();
+            console.log('Timed bounty expiry stored:', result);
```
[the remark this answers is not in the document]
Comment (Nachiket-Roy) [reply .github/workflows/timed-bounty-label.yml]: @coderabbitai in this scenario indexing is needed?

**Review by Nachiket-Roy — APPROVED** [review]
Thanks - all my concerns are addressed..

**Comment by Nachiket-Roy** [comment]
Also look at the last Nitpick comments [2] if you'd like.

**Review by yashhzd — COMMENTED** [review]
Reviewed the timed bounty implementation across workflows, views, models, and tests. The two-transaction payment pattern is a smart design choice. Feedback below:

## What works well

1. **Two-transaction payment flow** — Validation + `payment_pending` flag in one atomic block, external payment call outside any transaction, then recording `sponsors_tx_id` in a second atomic block. This avoids holding DB locks during the external GitHub Sponsors API call while still preventing duplicate payments.

2. **Rate limiting decorator** — The `webhook_rate_limit` using `cache.incr()` with `cache.add()` fallback is clean. Failing open on cache errors is the right call for a webhook endpoint. Using `REMOTE_ADDR` instead of `X-Forwarded-For` avoids header spoofing, and the comment about django-ipware for proxied setups is helpful.

3. **Bounty label parsing** — `parseBountyLabel()` handles both timed (`$5 24`) and regular (`$50`) labels cleanly. The `MAX_DURATION_HOURS = 720` cap (30 days) is validated in both the GH Action and the Django endpoint, which is good defense-in-depth.

4. **Expiry validation** — Both past-date and over-max-duration checks with clear error messages. Naive datetime conversion to UTC via `timezone.make_aware()`.

5. **Tests** — Good coverage of the critical paths: timed bounty recording, invalid datetime rejection, expired bounty rejection, missing expiry with `is_timed_bounty=True`, and the normal non-timed flow.

## Potential concerns

1. **f-strings in logger calls** — Several places use `logger.warning(f"...")` and `logger.error(f"...")`. This eagerly evaluates the format string even when the log level is disabled. Consider using lazy formatting: `logger.warning("Message %s", var)` instead. This is a minor performance point but aligns with Django logging best practices.

2. **`hasattr(github_issue, "payment_pending")` guards** — After the migration runs, `payment_pending` always exists on `GitHubIssue`. These `hasattr` checks suggest the code needs to work pre-migration, but that's fragile since the field has `default=False` and Django handles it. If you want pre-migration safety, a try/except on `FieldDoesNotExist` would be more explicit.

3. **Stale `payment_pending` recovery** — If the external payment call fails AND the cleanup transaction also fails (line ~305), the issue stays `payment_pending=True` permanently with no recovery path. Consider adding a TTL-based cleanup (e.g., a management command that clears `payment_pending` flags older than 1 hour) or resetting in the `timed_bounty` endpoint when recording a new expiry.

4. **Bounty label regex requires exactly 2 decimal places** — `^\$(\d+(?:\.\d{2})?)\s+(\d+)$` matches `$5 24` and `$25.50 24` but rejects `$5.5 24`. This might confuse users creating labels. Consider `\.\d{1,2}` to accept both `$5.5` and `$5.50`.

5. **`permissions: {}` at workflow top level** — The empty permissions block in `timed-bounty-label.yml` drops all default permissions, which is good for security (principle of least privilege). Just making sure the job-level `contents: read` and `issues: read` are sufficient — might also need `issues: write` if you plan to add auto-commenting on label events later.

Overall well-structured implementation with proper atomicity guarantees. The `payment_pending` pattern is a solid approach to prevent double-spending.

**Comment by Iceshen87** [comment]
Hi OWASP-BLT team! 👋

**I can help with the timed bounty action feature!**

**My Plan:**
1. Review current timed bounty implementation
2. Test the new action flow
3. Fix any issues found
4. Add tests if needed
5. Update documentation

**Timeline:** 2-3 days
**Bounty:** Open to discussion

**Why I am a good fit:**
- Experience with GitHub Actions
- Familiar with bounty/automation systems
- Strong CI/CD background

Ready to help!

- Ethan (Iceshen87)

**Review by DonnieBLT — DISMISSED** [review]
Please simplify this into smaller steps.

**Review by yashhzd — APPROVED** [review]
Thorough implementation of timed bounties. The overall architecture is solid — the label webhook records expiry, and the payout webhook validates it before processing. A few observations:

**Strengths:**
- `payment_pending` flag with `select_for_update()` prevents duplicate payments — good concurrency handling
- Rate limiting decorator with fail-open behavior is well-designed
- Comprehensive test coverage (12 new tests covering expiry, invalid dates, stale cleanup, pending states, etc.)
- `secrets.compare_digest` for token comparison prevents timing attacks
- Cleanup retries with exponential backoff on payment failure is robust
- Clearing stale `bounty_expiry_date` when label changes from timed to non-timed is a smart edge case to handle

**Suggestions:**
1. **`time.sleep()` in view code** (line ~320 in bounty.py) — the cleanup retry loop uses `time.sleep(0.5 * attempt)` which blocks the request thread. For a webhook endpoint this is likely acceptable (caller is GitHub Actions, not a user), but worth noting.

2. **Double lookup in `timed_bounty`** — the issue is fetched with `.filter().first()` at line ~145, then fetched again with `select_for_update().get()` inside the transaction at line ~165. The first fetch could be removed since the second one handles `DoesNotExist`. This would reduce one unnecessary query.

3. **`MAX_DURATION_HOURS` is duplicated** — defined as 720 in both `timed-bounty-label.yml` (workflow) and `timed_bounty()` (view). Consider making the view the single source of truth and documenting the constant.

4. **Migration dependency** — the migration depends on `0270_githubcomment`. Make sure this is still the correct leaf migration after merging upstream main, as migration conflicts are common in active repos.

Good work overall — this is a well-tested feature with solid error handling.

**Comment by DonnieBLT** [comment]
please move this to blt-rewards
